lib/pbench/client/oidc_admin.py

```diff
@@ -0,0 +1,150 @@
+from http import HTTPStatus
+import json
+import os
+
+import requests
+
+from pbench.server.auth import Connection
+
+
+class OIDCAdmin(Connection):
+    OIDC_REALM = "pbench-server"
+    ADMIN_USERNAME = os.getenv("ADMIN_USERNAME", "admin")
+    ADMIN_PASSWORD = os.getenv("ADMIN_PASSWORD", "admin")
+
+    def __init__(self, server_url: str):
+        super().__init__(server_url, verify=False)
+
+    def get_admin_token(self) -> dict:
+        """pbench-server realm admin user login.
+
+        Returns:
+            access_token json payload
+
+            { 'access_token': <access_token>,
+              'expires_in': 60, 'refresh_expires_in': 1800,
+              'refresh_token': <refresh_token>,
+              'token_type': 'Bearer',
+              'not-before-policy': 0,
+              'session_state': '8f558797-50e7-496d-bb45-3b5ac9fdcddb',
+              'scope': 'profile email'}
+
+        """
+        url_path = "/realms/master/protocol/openid-connect/token"
+        data = {
+            "grant_type": "password",
+            "client_id": "admin-cli",
+            "username": self.ADMIN_USERNAME,
+            "password": self.ADMIN_PASSWORD,
+        }
+        return self.post(path=url_path, data=data).json()
+
+    def create_new_user(
+        self,
+        username: str,
+        email: str,
+        password: str,
+        first_name: str = "",
+        last_name: str = "",
+    ) -> requests.Response:
+        """Creates a new user under the OIDC_REALM.
+
+        Note: This involves a REST API call to the
+        OIDC server to create a new user.
+
+        Args:
+            username: username to register,
+            email: user email address,
+            password: user password,
+            first_name: Optional first name of the user,
+            last_name: Optional first name of the user,
+
+        Returns:
+            Response from the request.
+        """
+        admin_token = self.get_admin_token().get("access_token")
+        url_path = f"/admin/realms/{self.OIDC_REALM}/users"
+        headers = {
+            "Content-Type": "application/json",
+            "Authorization": f"Bearer {admin_token}",
+        }
+        data = {
+            "username": username,
+            "email": email,
+            "emailVerified": True,
+            "enabled": True,
+            "firstName": first_name,
+            "lastName": last_name,
+            "credentials": [
+                {"type": "password", "value": password, "temporary": False}
+            ],
+        }
+        response = self.post(path=url_path, data=json.dumps(data), headers=headers)
```

Comment:
Line 68 says we're going to send a JSON object; line 71 builds a nice dictionary; and then line 82 converts it to a _string_?  Sending `data=data` doesn't work?  (It seems to work at lines 40 and 114....)

Reply:
It's because of the Python requests package and the fact that API on this Keycloak endpoint requires JSON data. If I just send `data=data` the Keycloak API doesn't accept it as JSON and totally ignores the header `"Content-Type": "application/json"`. However, if I use `json` parameter of the `request` module like `json=data` then the endpoint correctly interpret the data as json format (with `json` parameter `application/json` automatically gets inserted as a header).   

Our Connection class only accept the `data` parameter so I had to serialize the dict to a JSON formatted str and send it as data value with `application/json` header for it to work.

Reply:
> If I just send `data=data` the Keycloak API doesn't accept it as JSON and totally ignores the header `"Content-Type": "application/json"`.

I don't know who to blame for that, but...

> if I use `json` parameter of the [request](https://docs.python-requests.org/en/latest/api/#requests.Request) module like `json=data` then the endpoint correctly interpret the data as json format (with `json` parameter `application/json` automatically gets inserted as a header).

Given what you say here, it sounds like we're supposed to be using `json` rather than `data` for a JSON payload.

Given _that_, I find it interesting that you opted to expand the `post()` interface to allow it to accept a `str` value for the `data` parameter rather than adding a `json` parameter....  😀

Reply:
Yeah, now that you mentioned it, I realize I should have done that. I'll fix it, anyway I need to fix some unit test failures.

Reply:
Cool.  And, now that you've brought this to light, I'm wondering what other places are "a little off"....

The docs for Requests indicate that if you specify a `dict` to `data` that "form-encoding takes place", whatever that means, but it certainly doesn't result in `"Content-Type": "application/json"`....

Reply:
> "form-encoding takes place", whatever that means

I think it means as though you were submitting a _form_ on an HTML page

Reply:
Right...it means `application/x-www-form-urlencoded` where we presumably want `application/json`....